### PR TITLE
NS2.1 Sync new changes, mobile profiles

### PR DESCRIPTION
**Pull request by EgorBo**
Sync the latest ns changes with Mono (Index, Range, IAsyncDisposable, etc). Also some progress on mobile profiles (almost finished `monodroid`).

### CODE REVIEW

File: mcs/class/Facades/netstandard/TypeForwarders.cs [hunk 1]
```diff
@@ -2433,14 +2443,16 @@
 [assembly:System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Buffers.Text.Utf8Parser))]
 [assembly:System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Collections.Generic.CollectionExtensions))]
 [assembly:System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Collections.Generic.KeyValuePair))]
-[assembly:System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Data.Common.DbProviderFactories))]
 [assembly:System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Diagnostics.Tracing.EventSourceCreatedEventArgs))]
-[assembly:System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Drawing.ColorConverter))]
 [assembly:System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Drawing.KnownColor))]
+[assembly:System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Data.Common.DbProviderFactories))]
+#if !MOBILE
```
Comment (marek-safar) [inline mcs/class/Facades/netstandard/TypeForwarders.cs]: This needs to be in mobile as well, @akoeplinger could you talk to @EgorBo how to do it

File: mcs/class/Facades/netstandard/TypeForwarders.cs [hunk 2]
```diff
@@ -2527,6 +2539,9 @@
 [assembly:System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.SequencePosition))]
 [assembly:System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Security.Cryptography.CryptographicOperations))]
 [assembly:System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Security.Cryptography.ECDiffieHellman))]
+#if !MOBILE
+[assembly:System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Xml.Serialization.SchemaImporter))]
```
Comment (marek-safar) [inline mcs/class/Facades/netstandard/TypeForwarders.cs]: Needs to be in mobile as well (could be PNSE)

File: mcs/class/Facades/netstandard/TypeForwarders.cs (hunk 1 above)
Comment (EgorBo) [reply mcs/class/Facades/netstandard/TypeForwarders.cs]: @marek-safar yeah I know, I wrapped it with #if-def temporarily for a green build 🙂

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by EgorBo** [comment]
Current state by profiles:
```
net_4_x:
Total Issues: 0
monodroid:
Total Issues: 0
xammac:
Compat issues with assembly netstandard:
TypesMustExist : Type 'System.Drawing.Color' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.ColorConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.KnownColor' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.Point' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.PointConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.PointF' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.Rectangle' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.RectangleConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.RectangleF' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.Size' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.SizeConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.SizeF' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.SizeFConverter' does not exist in the implementation but it does exist in the contract.
Total Issues: 13
monotouch:
Compat issues with assembly netstandard:
TypesMustExist : Type 'System.Drawing.Color' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.ColorConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.KnownColor' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.Point' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.PointConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.PointF' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.Rectangle' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.RectangleConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.RectangleF' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.Size' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.SizeConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.SizeF' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.SizeFConverter' does not exist in the implementation but it does exist in the contract.
Total Issues: 13
monotouch_watch:
Compat issues with assembly netstandard:
TypesMustExist : Type 'System.Drawing.Color' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.ColorConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.KnownColor' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.Point' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.PointConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.PointF' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.Rectangle' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.RectangleConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.RectangleF' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.Size' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.SizeConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.SizeF' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.SizeFConverter' does not exist in the implementation but it does exist in the contract.
Total Issues: 13
monotouch_tv:
Compat issues with assembly netstandard:
TypesMustExist : Type 'System.Drawing.Color' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.ColorConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.KnownColor' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.Point' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.PointConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.PointF' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.Rectangle' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.RectangleConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.RectangleF' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.Size' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.SizeConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.SizeF' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.SizeFConverter' does not exist in the implementation but it does exist in the contract.
Total Issues: 13
wasm:
Compat issues with assembly netstandard:
TypesMustExist : Type 'System.Drawing.ColorConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.PointConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.RectangleConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.SizeConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.SizeFConverter' does not exist in the implementation but it does exist in the contract.
Total Issues: 5
winaot:
Compat issues with assembly netstandard:
TypesMustExist : Type 'System.Drawing.ColorConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.PointConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.RectangleConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.SizeConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.SizeFConverter' does not exist in the implementation but it does exist in the contract.
Total Issues: 5
xammac_net_4_5:
Compat issues with assembly netstandard:
TypesMustExist : Type 'System.Drawing.ColorConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.PointConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.RectangleConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.SizeConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.SizeFConverter' does not exist in the implementation but it does exist in the contract.
Total Issues: 5
unreal:
Compat issues with assembly netstandard:
TypesMustExist : Type 'System.Drawing.ColorConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.PointConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.RectangleConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.SizeConverter' does not exist in the implementation but it does exist in the contract.
TypesMustExist : Type 'System.Drawing.SizeFConverter' does not exist in the implementation but it does exist in the contract.
Total Issues: 5
```

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by marek-safar** [comment]
@EgorBo is decimalremainder failure related?

**Comment by akoeplinger** [comment]
@marek-safar there's a PR to fix it: https://github.com/mono/coreclr/pull/3 (note that the file is committed in UTF16 encoding which makes GitHub not show the diff)

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by marek-safar** [comment]
@monojenkins commit csproj and apidiff

**Comment by NattyNarwhal** [comment]
Regression: decimal changes look like they broke BE/aix, since the error message looked familar, I had a peek. turns out they basically rewrote the thing *AND* undid the changes I made to check for endianess at runtime instead of at compile-time.

**Comment by marek-safar** [comment]
ouch, @EgorBo could you diff the changes?

**Comment by EgorBo** [comment]
@marek-safar @NattyNarwhal oops, sorry totally forgot about it - will try to fix and upstream it

**Pull request merged**